### PR TITLE
Add comparison between AppSync and Sangria for GraphQL

### PR DESCRIPTION
**Pull request by suzannehamilton**
Write up our investigations into GraphQL servers, comparing AWS AppSync and the Sangria Scala library.

This is a draft PR until we add some notes on why it was hard to connect to the DB from AppSync. @ian-hoyle or @MancunianSam - could you help me out with that?

### CODE REVIEW

File: technology-considerations/graphql-prototypes.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+### Local development
+
+It was fairly straightforward to set up an [AppSync
+prototype][tdr-appsync-prototype] based on the [Serverless
+Framework][serverless]. This connected to DynamoDB and just used Velocity to map
+the data.
+
+As mentioned above, we wanted to check that we could write business logic in
+Scala. [Serverless supports Lambdas written in Java][serverless-java] (which
```
Comment (ian-hoyle) [inline technology-considerations/graphql-prototypes.md]: IntelliJ https://plugins.jetbrains.com/plugin/11349-aws-toolkit can be used. https://docs.aws.amazon.com/toolkit-for-jetbrains/latest/userguide/key-tasks.html#key-tasks-install 
Requires 2018.3 or later so need licence extension

File: technology-considerations/graphql-prototypes.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+[AWS AppSync][appsync] is a platform specifically for building GraphQL APIs, and
+it can fetch data from multiple data sources including AWS RDS databases.
+
+In AppSync, [GraphQL resolvers][resolvers] can be implemented in one of two
+ways. Simple resolvers, which are just a wrapper for a single database table,
+can use [Apache Velocity templates][velocity] to map between the GraphQL request
+and the database. More complex resolvers, which implement business or
+authorization logic, can be written as [pipeline resolvers][pipeline-resolvers],
+which call arbitrary AWS Lambda functions.
+
```
Comment (ian-hoyle) [inline technology-considerations/graphql-prototypes.md]: Simple resolvers require an AppSync data source
 Aurora (a database that we are considering) is not a supported Data Source, lambda resolvers are required.
 Aurora Serverless can be used as a data source when set up to use the Data API. Data API is not available in the London region.
  To use Aurora Serverless without the Data API lambdas within a private VPC must be used and these can have a slow cold start.

File: technology-considerations/graphql-prototypes.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+### Hosting
+
+Using AppSync would make the application extremely coupled to AWS
+Infrastructure, because so much of the code would be in Velocity templates and
+the rest would be implemented as AWS Lambdas.
+
+### Database connection
+
+TODO: Ian or Sam to add details of why it was hard to connect to the DB
```
Comment (ian-hoyle) [inline technology-considerations/graphql-prototypes.md]: Simple resolvers require an AppSync data source
Aurora (a database that we are considering) is not a supported Data Source, lambda resolvers are required.
Aurora Serverless can be used as a data source when set up to use the Data API. Data API is not available in the London region.
To use Aurora Serverless without the Data API lambdas within a private VPC must be used and these can have a slow cold start.

File: technology-considerations/graphql-prototypes.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+haven't been able to find a good client library for Scala. The best option is
+[Drunk], but it appears to be a very thin wrapper around an HTTP client and
+the Circe JSON parser. You have to write graphQL queries as strings, with no
+type safety to protect you from syntax errors or inconsistencies between the
+code and the schema.
+
+For now, we've chosen to just use an HTTP client library directly. This gives
+us more control over the HTTP request, which is particularly useful when we need
+to do things like [sign requests][iam-signing].
+
```
Comment (ian-hoyle) [inline technology-considerations/graphql-prototypes.md]: This [sbt plugin](https://github.com/muuki88/sbt-graphql) might 
allow generation of required classes from a schema.

File: technology-considerations/graphql-prototypes.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+This document is mostly about the choice of GraphQL server, but we've also
+learned a bit about GraphQL client libraries.
+
+Sangria has been easy to work with as a Scala GraphQL server library, but we
+haven't been able to find a good client library for Scala. The best option is
+[Drunk], but it appears to be a very thin wrapper around an HTTP client and
+the Circe JSON parser. You have to write graphQL queries as strings, with no
+type safety to protect you from syntax errors or inconsistencies between the
+code and the schema.
```
Comment (ian-hoyle) [inline technology-considerations/graphql-prototypes.md]: This sbt plugin might
allow generation of required classes from a schema.

File: technology-considerations/graphql-prototypes.md (hunk 1 above)
Comment (suzannehamilton) [reply technology-considerations/graphql-prototypes.md]: Do you know if it supports AppSync? I can't see any mention of it.

File: technology-considerations/graphql-prototypes.md (hunk 5 above)
Comment (suzannehamilton) [reply technology-considerations/graphql-prototypes.md]: I've got schema generation working in https://github.com/nationalarchives/tdr-prototype-sangria/pull/11. Do you want to try it out in one of the clients?

File: technology-considerations/graphql-prototypes.md (hunk 3 above)
Comment (suzannehamilton) [reply technology-considerations/graphql-prototypes.md]: Thanks. :+1: 

I've updated this section - can you check that it makes sense?

File: technology-considerations/graphql-prototypes.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+To connect AppSync to a relational DB such as Aurora, you can either use a [Data
+API][rds-data-api] or use a Lambda as a data source, both of which have
+downsides.
+
+[Data API is not currently available in eu-west-2][data-api-availability], and
+we need to store data in the UK.
+
+Lambda data sources suffer from cold starts, which may be even slower in a VPC
+(although [AWS say this will improve][lambda-vpc-improvements]).
+
```
Comment (MancunianSam) [inline technology-considerations/graphql-prototypes.md]: Amazon are rolling out a fix for the slow VPC cold starts. https://aws.amazon.com/blogs/compute/announcing-improved-vpc-networking-for-aws-lambda-functions/

File: technology-considerations/graphql-prototypes.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+The development process is simple because it's just a single Scala application.
+We've wrapped the Sangria code in an [akka-http] server for development, which
+is quick to redeploy when modifying the code.
+
+We've added several API endpoints so far, and it's been fairly straightforward
+to build everything we need.
+
+### Deployment and hosting
+
```
Comment (MancunianSam) [inline technology-considerations/graphql-prototypes.md]: We're not running this in a lambda any more so the akka-http server is what we're using in the dev environment on ECS as well as for local development and we're not using the lambda package at all.

File: technology-considerations/graphql-prototypes.md (hunk 6 above)
Comment (suzannehamilton) [reply technology-considerations/graphql-prototypes.md]: I've added that link, thanks. :+1:

File: technology-considerations/graphql-prototypes.md (hunk 7 above)
Comment (suzannehamilton) [reply technology-considerations/graphql-prototypes.md]: I've updated this section.

File: technology-considerations/graphql-prototypes.md (hunk 1 above)
Comment (ian-hoyle) [reply technology-considerations/graphql-prototypes.md]: Don't know if it supports appsync. Just  AWS toolkit could be used for our Lambda development

File: technology-considerations/graphql-prototypes.md (hunk 3 above)
Comment (ian-hoyle) [reply technology-considerations/graphql-prototypes.md]: yes

File: technology-considerations/graphql-prototypes.md (hunk 5 above)
Comment (ian-hoyle) [reply technology-considerations/graphql-prototypes.md]: updated to use modified version of Drunk with code generated from graphql schema and files

Comment (suzannehamilton) [reply technology-considerations/graphql-prototypes.md]: Cheers. I've fixed the git branch issue and added a bit about adapting Drunk.

**Comment by suzannehamilton** [comment]
@ian-hoyle @MancunianSam Happy for me to merge this?

**Pull request merged**